### PR TITLE
Add the ability to track the number of requests issued

### PR DESCRIPTION
**Pull request by MakersF**
This PR adds the ability (disabled by default) of tracking the requests which are made.

This allows users of the library to track of the usage of their limit, allowing to better investigate performances and efficiency of their code.

I find this fundamental as I try a different architectural approach for my library, to be able to compare if I'm effectively improving the performances or not.

This would make debugging issues like #16 a lot easier as well.

Accept one PR, get one fix for free: fixed a typo in the documentation

### CODE REVIEW

**Comment by jjmaldonis** [comment]
I think this is a good addition to the library and something I want to use. A few comments:

I personally like the name "count_requests" slightly better because I think it is more descriptive the the exact problem. However, I think this should be default and that the user shouldn't turn it on/off. The over head is minor and it won't be printing anything to screen, so it's hidden from the user completely unless they access it (and the overhead is trivial).

I think "get_requests_count" should be  a property of dto/requests if at all possible and that the getter should be removed.  I don't think you can have a property in a module unfortunately, but I think that just making that variable accessible to user is fine. If they change it, well... that's on them. Python doesn't have getters though. We might therefore need a better name for that variable and need some way to document it.

Why are you using threading? Is this the only point in the library where a thread lock is necessary? I doubt it (although I am unsure), so it should be an all-or-nothing thing imo. Either everything get's thread locked, or nothing does.

Do we really want to track retries as extra requests? Maybe we should track them separately as "number of requests" and "number of retries". Otherwise I think it could really throw someone off, mostly because I don't think re-tries are printed (not 100% sure though). I think moving your implementation to be between 72-73 makes more sense personally.

Rather than having an assert statement in baseriotapi, you should raise the Cass exception. Raising an assertion error is not clear, and there are better ways to do that (i.e. if people are doing error handling, they are not going to expect an assertion error).

I don't think there is another way to increment the counter without using `global`, but I wish there was personally. `globals['requests_number'] += 1` is another way, but I don't like it any better.

Lastly, fixing a doc typo should not be a reason to accept this PR. Thanks for the catch though.

**Comment by MakersF** [comment]
Ahah, of course fixing a typo is not a reason to accept such a PR. I just
did it because it was in the doc just above the function I added, and I was
basically copying the structure of the doc, so I noticed it.

Name change to "count_requests": makes sense. I'll update the PR

Overhead is minor: maybe. Actually with a production key and an heavily
multithreaded code skipping a locking step that you don't need might have
some impact. We should check on that (now it's possible as we can count the
requests :D )

About the getter: I think the point of the library is having all the
functionality in riotapi.anymethod and baseriotapi.anymethod. so having the
getting there is the consistent choice with the rest of the library. I
believe it to be cumbersome for people to have in their imports
import cassiopeia
import cassiopeia.dto.core.requests
The latter is not necessary now, but would if we only allowed access to the
module variable

Threading: I'm using a lock because that code is run inside the rate
limiter, which is multithreaded. Unfortunately python has no atomic
integers, so I need to lock to guarantee the correctness of the counter.
Removing the lock will bring missed updates

Extra requests: this counter has a really simple and well defined semantic:
"if riot servers receive a request, that is tracked on the counter". I
don't believe making retries separated would be an useful information. What
it tells you is how unreliable is the network/riot server, both of which
you don't have control over.
Even once you know you are doing 10% retries, what are you going to do with
that info?
I believe retries are printed, as the printing code is just a few lines
before the counting code. Should check the code to be sure though (I'm on
the phone now).

Assertion: absolutely true. I'll update the PR.

I don't like the global as well, but either we use a class with an
attribute that we use as a counter, or we use that. I didn't think that
adding a " dummy" class with only the attribute "count" would have been
clear or good. It makes the code a bit more messy. I think we should leave
it as it is now and maybe change it later if a different consensus is
reached.

**Comment by jjmaldonis** [comment]
> Overhead is minor: maybe.
A single request checks the rate limiter, makes the request to riot (which takes noticeable time), then parses the json into an object. In comparison to that, hopefully even a thread lock is very minor. I don't know for sure, but will you check? --- Actually, read the below. If you agree, this won't be necessary.

It might be possible to avoid an additional lock altogether, however. The SingleRateLimiter class already tracks how many calls are currently being executed (and it locks too). It would be simple to add a new `self.total` value and increment it on every call (it should probably be a `property` that references `self._total` though). Then the MultiRateLimiter could have a `property` called `total` that returns the sum of the `total`'s in each of it's SingleRateLimiters. This avoids a new global variable that needs to be set in dto/requests.py. Instead, that variable would be accessed through `dto.requests.rate_limiter.total` (maybe using a getter).

I checked and failed requests that are retried _are not_ printed to screen. Here is a use-case for why I think keeping track of successful and unsuccessful requests is important:  What if am new and I don't really know what I am doing and I want to test how many requests are made during both eager loading and lazy loading (so I can pick which one I want to use)? I would run my program twice, once with eager loading and once with lazy loading, and print the total number of calls. If it's a close call, failed and retried requests will affect the outputs and I won't be sure which implementation is best.

I would argue that including successful and unsuccessful requests in one counter tells you how unreliable the network / Riot's server is more so than counting both the number of successful calls and the number of unsuccessful ones separately. By the way, this implementation also has a well defined semantic:  "count the number of successful calls you made to Riot's servers, which is the number of calls that have an effect on your code". This value should then be 100% reproducible every time your code is run, and it will be independent of finicky server problems.

I'm still not sure about the getter in riotapi. Python's ideology is that something should only be a function if it actually does something. If all you are doing is accessing a value, you should just look at the value, not call a getter function to look at it for you.

Now that I think about my comments about the rate limiters above, we could modify riotapi.set_rate_limit to return the rate limiter (i.e. the user would do `rate_limiter = riotapi.set_rate_limit((3000, 10))`) and then you could access the total number of calls by `rate_limiter.total` (or maybe `rate_limiter.total_calls`). _That_ actually sounds like good python implementation to me. The variable is contained in a stateful environment (a class) and it's protected from user modification because it's a property. Plus, it's easily accessible when the rate limit is set.

So all in all, putting a `total_calls` property in the RateLimiter classes 1) gets around the issue of making another lock (no extra overhead; don't need to worry about threading again), 2) removes the need for a getter in riotapi, 3) removes the need for a new global variable in dto/requests.py, and 4) allows that value to be protected via a property so the user can't modify it unless they really really want to.

I am not sure if we will be able to track successful requests vs. unsuccessful requests or not if things are moved into the rate  limiters (but I hope my above comments have convinced why I think it's important to have counters for both). I'll need to think on this some more. Maybe you'll figure it out though.

Let me know what you think. This should come together reasonably well in the end I believe.

**Comment by MakersF** [comment]
> I checked and failed requests that are retried are not printed to screen

Really? From the code it seams that the print happens before the request and if the request raises an error it is reissued. So I'd expect the print to happen again.
Moreover the code update the counter in the same function where it prints the call, so shouldn't it be behaving the same way?
I must be missing something..


> Here is a use-case

I see your point and I changed my mind.
We should only count the "real" requests without the retries.
The reason is that the Riot Servers and the network are external variables. When an user performs the comparison for performance the servers or the network might behave differently from how they normally behave. And the user network is not representative of the quality of the network of very user of its library. BUT we may expect that the failed rate is a percentage of the issued requests.
So if an approach issues less "real" requests, there will also less retries as well (because the server/the network drop a % of them, independently from the approach used).
I'm still of the opinion that retries are not an useful information and we shouldn't add complexity to address it.


> This value should then be 100% reproducible every time your code is run, and it will be independent of finicky server problems.

This is indeed a nice, deterministic property to have.


> I'm still not sure about the getter in riotapi. Python's ideology is that something should only be a function if it actually does something. If all you are doing is accessing a value, you should just look at the value, not call a getter function to look at it for you.

Again, I see your point. It makes sense.
The pitfall I see is this one
`from cassiopeia.dto.requests import requests_count`
Now requests_count will always stay at 0, since it's a reference to an immutable int.
Should be instead `from cassiopeia.dto import requests` and use it as `requests.requests_count`.
If we used the class instead this problem wouldn't arise.

> [Everything else about threading and rate limiters]

My first idea was to write it as a decorator of the limiter which you could use if you wanted to count the requests. The reasons are similar to yours: the lock is already there.
But then if I wanted to use the lock already present I couldn't use a decorator and I should have modified the rates.
I'm not sure that was the best thing to do.
The code I committed here is simpler and more self contained, and I had no idea if you'd like this kind of feature.

If we move it to the rate limiter we will be able to do a count like now if we use the locks already in place, and I think we should be able to count only the successful requests if we add a new acquire/release.

**Comment by MakersF** [comment]
This should address most of the points.

Now the function baseriotapi makes sense because it switches between tournament and not tournament calls.

Differentiating retries vs not retries might become involved, I'm not really sure if it's worth the amount of work needed to do that (I think we can do that by moving the counting again into requests and adding a flag to track down if the current request is "new" or not)

**Comment by jjmaldonis** [comment]
>From the code it seams that the print happens before the request and if the request raises an error it is reissued. So I'd expect the print to happen again.

I think you're right, I missed the `return get(request, params, static)` on line 79. When I tested the code,  I must not have gotten a 429 like I thought; the requests just were taking a really long time and I took that as Cass retrying. (Also I think there is another bug on line 79 -- it doesn't call all the default arguments from the original call. Would you mind fixing that `get` call to also pass in `include_base` and `tournament`? If I make a commit then it will cause an irritating merge conflict when Rob tries to accept this PR.)

>Differentiating retries vs not retries might become involved, I'm not really sure if it's worth the amount of work needed to do that.

That might be true. Still, we could add a new argument to the rate limiters' `.call` called `retry=False`, and only increment the counter if `retry` is True. That should be relatively simple and it should work.

>I'm still of the opinion that retries are not an useful information and we shouldn't add complexity to address it.

I'm good with this. If someone wants that functionality later, they can make an issue/PR.

>    @property
    def total_calls(self):
        """Returns the number of calls served by this limiter
        return    int    the number of calls served by this limiter
        """
        try:
            self.lock.acquire()
            return self.calls_count
        finally:
            self.lock.release()

What if we just `return self.total_calls` instead? This is just a read, but I think I see your point about the lock because it's possible that one thread is reading while another thread is writing. Still, hopefully that won't happen (and even if it does, technically the call hasn't "finished" so maybe it shouldn't be recorded yet anyways). I don't know much about threading, but I think that read statements don't usually have locks around them unless something special is going on. That's mostly a guess though.

In the MultiRateLimiter, what if we `return sum(limit.total_calls for limit in limits)`? That would eliminate the need for keeping track of the counters twice (the MultRateLimiter will still call the SingleRateLimiters via `limit.call` inside the `finally`). Then it isn't possible for the MultiRateLimiter and the SingleRateLimiters to disagree, which would be awkward. I'm also guessing that you included the extra `calls_count` variable in the MultiRateLimiter to avoid multiple locks in the SingleRateLimiters.

Will you change the name `count_calls` to `_total_calls`? This is a bit nit-picky, but there are two reasons for it. First, if a property is a read-only substitute for a variable (and it partly is here), Rob and I both use the syntax that the variable that the property refers to is the name of the property with an underscore before it. So any future changes in Cass that create a property will have that stylistic syntax, simply because that's how we do things. This will keep the library consistent. (You might note that many of the core types don't have this syntax, but that's because they refer to the dto data and are therefore a bit different in meaning). Secondly, variables that start with an underscore in python are implicitly "private", meaning that the user should know that they should only change them if they really know what's going on. This protects the variable. Right now, the user might read the code and see `SingleRateLimiter.count_calls` and just access it right from there. This would be fine, unless s/he modifies `count_calls` (which could be bad if future implementations rely on it). Naming it `_total_calls` makes it a little more "mysterious" and the user will hopefully understand that they shouldn't modify it unless they know exactly what the consequences are.

I'm still not sure about the getters because I think `return cassiopeia.dto.requests.rate_limiter` in `set_api_key` / `set_rate_limit` (and similarly for the tournament limiter) coupled with the ability to call `*RateLimiter.total_calls` provides similar functionality. It sounds like we will agree to disagree here though, so we can leave it as-is for now and get a second opinion from Rob. No worries.

Have you ever done a rebase? I haven't, but the commits are becoming overlapping enough that a final rebase at the end might be really useful. This might be worth a try... http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit

I think things are 90% done now. Thanks for your work, this will be a nice and useful addition imo. At the very end we can make a nice summary for Rob and run it past him, and he can decide if there is anything that he doesn't like (I think it'll be good though), and then accept.

**Comment by MakersF** [comment]
> Would you mind fixing that get call to also pass in include_base and tournament?

I'd rather not push external fixes to this PR. If you want to submit a PR for that do so, I can merge. Right now I'm not touching the requests file, so I wouldn't even have to merge.
That fix can go in way faster than this Pr, which seems like it's taking a bit of time

> we could add a new argument  ...

It's doable. Let's hear from Rob if he thinks it's ok.

> What if we just return self.total_calls instead?

Depending on the memory model of Python (which I believe being similar to Java's one), not locking would not synchronize the variable, which means that it could use an old value in cache, being severely older than the real one.
I believe that in practice it might work just fine(the read is atomic so there are no concurrency problems).
I'm not really sure about committing unsafe code though. I think it's worth keeping the lock.
I don't think you should worry too much on the performances on that call, you are not going to call it on tight loops.

> In the MultiRateLimiter, what if we ..

We can't. MultiRateLimiter calls `call` on every limiter every time (to keep them synchronized). This means that `all(limit.total_calls == k for limit in limits)` is true. So we could just use `self.limits[0].total_calls`. But what if there are no limits (MultiRateLimiter behaves like a normal call to function in that case)?
My solution cover any possible case of usage of MultiRateLimiter.

> Will you change the name count_calls to _total_calls?

Sure

> Have you ever done a rebase?

When it's ready to go I can squash all the commits in only one and merge that one so we don't have a crazy long history.


To recap, I'll rename the variable and then let's wait for Rob before I go on doing work which might not be in the right direction

**Comment by MakersF** [comment]
I can't find any documentation about python memory model and stale reads. If you find anything please like it here so that we know if it's safe to remove the lock for the read

**Comment by jjmaldonis** [comment]
One of my goals with Cass is to help Rob not have to read a bunch of commits and 3 pages of comments. Unfortunately he's gotten a ton of pull requests for Orianna lately in addition to Cass, and he's working on implementing the tournament API and other things. I'm less busy right now, so I'm help  taking care of the Cass pull requests until they are essentially ready to be merged. With the goal of not making him read 7 commits (some of which undo each other) and 3 pages of comments, please squash things when you finish this next commit. It will be much easier for him to understand or me to explain. If you could also point out all the main design choices we have been discussing inside the commit message, that would be great.

>It's doable. Let's hear from Rob if he thinks it's ok.

Imo it's easier to see an implementation and decide if it's a good idea than to try to imagine the implementation and decide if it's a good idea. For the sake of clarity, please add the `retry` parameter to your next commit so Rob can clearly see what we are talking about (and feel free to put in a comment saying you don't like it if that's the case; a comment explaining why it's there would be helpful too).

>So we could just use self.limits[0].total_calls.

This is my preferred implementation for the reasons I have discussed before.

>But what if there are no limits (MultiRateLimiter behaves like a normal call to function in that case)? My solution cover any possible case of usage of MultiRateLimiter.

What is the use case for having no SingleRateLimiters at all? That is effectively disabling a major feature of this library (rate limiting). If someone wants to make that significant of a change (even though they can do it using functions we provide), I feel like they can modify it in their own fork. Rate limiting is one of the main features of Cass, and I don't think we should design the library with implementations that purposely allow for weird use cases (like disabling rate limiting).

I realize that it's just an extra counter in MultiRateLimiter, but it's duplicate functionality and maybe there is some weird scenario where the MultiRateLimiter and the SingleRateLimiter would disagree. I don't think this is a huge deal, but please mention it in your squashed commit message if you still don't want to change it.

I did some reading on reads / writes and locking (there wasn't a good article that I found either). It's something I'm not very familiar with, but from what I can tell the lock on the read is necessary, so I'm happy with leaving the property as-is. And like you said, it's a pretty minor speed thing because it is only one lock and won't be used in a loop. That's no big deal.

When you have made the above changes, I'll let Rob know that we are ready for him to take a look.

Also, if you want to email me your email address I'll add you to our slack group.

**Comment by MakersF** [comment]
Rebased to merge with your fix and squashed everything in 2 commits:
the first one does all the things we have agreed upon.
The second adds the not-counting-retries feature (which I'm don't fully agree to, as it doesn't look really solid).

The commit message and diff should be enough to have a clear picture, it's just 3 files (the other 2 file changes are for the tests).
I'd really not keep doing and undoing the counting now until it is decided the correct way in case this one needs change (this is the 3rd implementation now)

I also sent you an email, @jjmaldonis

**Comment by MakersF** [comment]
I'd like to add some notes about the _sentinel value.
A starting point is https://en.wikipedia.org/wiki/Sentinel_value and line
40 of rates where I specify what it is for.

Originally, the multi limiter directly invoked the passed function after it
waited for all the single limiters, then called the single limiters with no
function so that they updated their timers.

Now we want to distinguish if the function raises an exception. It's easily
done in the single limiter, the problem is in the multirate limiter.
Since the multirate limiter is calling the function, but we are keeping the
calls count in the single limiters, we need a way to notify the single
limiter that he should update the successful calls only if the method
didn't throw.

A possible way is to pass a function which always throws to the single
limiters instead of `limiter.call()` with no args, so that they update the
correct counters.

Alternatively, we can use a special value to specify that the "parent"
caller raised an exception.

I went the second route, as after benchmarking on my pc I sow that raising
and catching exception was 1 order of magnitude slower than checking with
the if, and the code was cleaner.

So instead of having 2 possible values as arguments to the call function
1) None : only update the timers
2) Function : update the timers and call the function

We now also have a _sentinel.
Sentinel is checked with `is`, so there is no way for a user to
accidentally use it, and it's special meaning is explained in the comment
at line 40.

This allows to keep the count in the single rates but still behaving
correctly when the multirate calls a function which throws
